Add reset_on_copy adapter

sherm1 opened the pull request:
Adds `reset_on_copy<T>` adapter for data members that need to be reset after copy construction, typically counters and pointers that would be stale after a copy. This allows easy-to-maintain default copy constructors to be used more often. I'm using this in caching but expect it to be generally useful.

The code is modeled on `reinit_after_move<T>` (thanks, Jeremy).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8208)
<!-- Reviewable:end -->

SeanCurtis-TRI commented:
It seems odd that a copy (via assignment) doesn't reinit. That seems like counter-intuitive behavior. And this seems most apparent when `T` is not POD. The documentation doesn't allude to the distinction between one form of a copy and the other. I note that the source (`reinit_after_move.h`) implements *both* move semantics.

Generally, the relationship of this and `T` is a class/struct makes the whole thing hazy. Would it make sense to limit it to numerical and pointer types?

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[common/reinit_on_copy.h, line 11 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6TZe_Q9kLZ0qoA8jhc:-L6TZe_R7lMwL_QUMXeN:binnjz8) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/reinit_on_copy.h#L11)):*
I'd recommend borrowing more from the url you've linked below on the constructor documentation, e.g.:

> Note that value initialization means the initialization performed when a variable is constructed with an empty initializer. For example, numeric types are set to zero and pointer types are set to nullptr.

This leaves the door slightly more ajar to predict what happens when `T` is *not* POD.

---

*[common/reinit_on_copy.h, line 76 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6TaNweAZ10jGtC0015:-L6TaNwfDRqbsRAhyxbQ:bi6nc6l) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/reinit_on_copy.h#L76)):*
> ```objc
>   //@}
> 
>   /// @name Dereference operators available if type T is a pointer.
> ```

FYI Seems `reinit_after_move` would benefit from this as well. At the very least, it would make the two otherwise parallel APIs align better.

---

*[common/test/reinit_on_copy_test.cc, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Tatf3MRn7Q9OQuqSx:-L6Tatf3MRn7Q9OQuqSy:b-7s3hqh) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L13)):*
> ```cpp
> // it, EXPECT_EQ is underspecified whether we are unwrapping the first argument
> // or converting the second.
> int ForceInt(int value) {
> ```

It seems to me that this trick is *absolutely* necessary if `T` is non-POD.

---

*[common/test/reinit_on_copy_test.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6TcRUd-VFyaqME6u92:-L6TcRUd-VFyaqME6u93:b-mi6ctt) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L27)):*
> ```cpp
>  private:
>   std::vector<int> items_;
>   reinit_on_copy<int> use_count_;
> ```

BTW Doing this:

`reinit_on_copy<int> use_count_{3};`

will not have the effect that it would suggest. Probably needs to be documented.

---

*[common/test/reinit_on_copy_test.cc, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6TaistWLjR0eXWs1oj:-L6TaistWLjR0eXWs1ok:b-ypmto1) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L29)):*
> ```cpp
>   reinit_on_copy<int> use_count_;
> };
> ```

BTW Adding a non-POD example might help illustrate some of the list intuitive aspects:

```C++
struct Bar {
  DRAKE_DEFAULT_COPY_AND_MOVE_AND_ASSIGN(Bar)
  Bar() = default;
  int i{10};
};

int& BarI(Bar& b) {
  return b.i;
}

GTEST_TEST(ReinitOnCopyTest, NonPodData) {
  reinit_on_copy<Bar> b;
  EXPECT_EQ(BarI(b), 10);
  BarI(b) = 15;
  EXPECT_EQ(BarI(b), 15);
  reinit_on_copy<Bar>b_cpy(b);
  // Note: the *contained* numerical values are value-initialized to the declared default.
  EXPECT_EQ(BarI(b_cpy), 10);
}
```

---

*[common/test/reinit_on_copy_test.cc, line 84 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Tblk2OF5vevt5BBYA:-L6Tblk2OF5vevt5BBYB:bul5x81) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L84)):*
> ```cpp
> 
>   EXPECT_EQ(my_foo_ptr->use_count(), 25);
>   EXPECT_EQ((*my_foo_ptr).use_count(), 25);
> ```

BTW Perhaps one more test?
`EXPECT_EQ(&*my_foo_ptr, &my_foo);`
The tests above could pass, even if `my_foo` were cloned in the construction. This additional test precludes the possibility of cloning.

---

*[common/test/reinit_on_copy_test.cc, line 107 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Tc9AszNH4ltr1VugU:-L6Tc9AszNH4ltr1VugV:bh55xle) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L107)):*
> ```cpp
>   w = x;
>   EXPECT_EQ(x, 3);
>   EXPECT_EQ(w, 3);
> ```

See previous note on why this is surprising.

---

*[common/test/reinit_on_copy_test.cc, line 155 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Td-7lcMv3el8WA9Bn:-L6Td-7lcMv3el8WA9Bo:bkon76t) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L155)):*
> ```cpp
> 
>   // Move-assignment and lack of aliasing.
>   reinit_on_copy<int> w{22};
> ```

FYI This value 22 is never used.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6T_ikaENGhAMv-No3K:b-c0zs66)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[common/BUILD.bazel, line 41 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6TZORqvk8S158YaCLu:-L6TZORqvk8S158YaCLv:b-5tdd90) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/BUILD.bazel#L41)):*
> ```Python
>         ":number_traits",
>         ":polynomial",
>         ":reinit_after_move",
> ```

BTW Please list the new item here.

---

*[common/reinit_on_copy.h, line 4 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6TZd4a2YLihJLVRN93:-L6TZd4a2YLihJLVRN94:b-3g7c6e) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/reinit_on_copy.h#L4)):*
> ```objc
> 
> #include <type_traits>
> #include <utility>
> ```

FYI Unused?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I hadn't thought about those angles when I read this, but now I'm also concerned.  Having thought about it for a few minutes with a scratchpad now, my intuition is that copy-ctor and copy-assign should be aligned, yes, so copy-assign should indeed reinitialize.  Since move vs copy is difficult to discern from certain kinds of call sites, move should also be the same and reinitialize.  Furthermore, since value-initialized is the "empty" state for the wrapped object, as with reinit_after_move the moved-from object should also be reinitialized.  So in short, I think we should value-initialize :all-the-things:.

---

Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-:-L6TfN9c11Ue3ChlCfLf:baqiu5l)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
*[common/test/reinit_on_copy_test.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6TcRUd-VFyaqME6u92:-L6TjMOm1TPaoqpLzaZR:ba3c55i) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L27)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Doing this:

`reinit_on_copy<int> use_count_{3};`

will not have the effect that it would suggest. Probably needs to be documented.
</blockquote></details>

@seancurtis-tri I agree -- I think I should rename this `reset_on_copy` instead (that's what I called the Simbody's more-elaborate [equivalent](https://simbody.github.io/simbody-3.6-doxygen/api/classSimTK_1_1ResetOnCopy.html#details)). Do you agree that's a better name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
I agree that copy assignment should reinit -- will fix. Jeremy I'm not sure exactly what you are saying. Move should definitely *not* reinitialize the destination (std::vector resize wouldn't work properly in that case, e.g.). Move *could* reinitialize the source so that reinit_on_copy would also perform the action of reinit_after_move but I don't see why that side effect would necessarily be good to add to this one-trick pony. One could write `reinit_after_move<reset_on_copy<int>> thing_` to pick up both effects (I think). If it were to do both things I think we'd have to call it something else. Can you clarify your thought, please?

---

Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-:-L6TkmtJ0vW0FMItlGBK:b-iryd3p)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[common/test/reinit_on_copy_test.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6TcRUd-VFyaqME6u92:-L6Tn2GqyTEoh3JC0UXO:b-46zpc9) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L27)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

@seancurtis-tri I agree -- I think I should rename this `reset_on_copy` instead (that's what I called the Simbody's more-elaborate [equivalent](https://simbody.github.io/simbody-3.6-doxygen/api/classSimTK_1_1ResetOnCopy.html#details)). Do you agree that's a better name?
</blockquote></details>

So, I feel those are two different issues:

1. I'm not persuaded that "reset" and "reinit" are any different in this regard.
2. In the example I gave above, `use_count_` does *not* get reset to 3, but to 0. That's the unexpected effect to which I was alluding.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
> Move should definitely not reinitialize the destination (std::vector resize wouldn't work properly in that case, e.g.).

Can you more clearly elaborate an example problem?  I couldn't come up with one, and the std::vector citation isn't enough to help me understand yet.

My reason for saying that move-ctor's recipient (the new object) should reinitialize its value, instead of moving-from the donor, is for something like this:
```
Foo foo(bar());
```
Depending on the signature of `bar()`, foo could be move-constructed or copy-constructed; we don't know unless we look at `bar()`, and of course that could easily change over time (and often will).  Given that ambiguity, it seems like all new `Foo` objects should always start out initialized, no?  In cases where someone wants to transfer an intermediate counter, they can write explicit `foo.value = ...` assignments.

> Move could reinitialize the source so that reinit_on_copy would also perform the action of reinit_after_move but I don't see why that side effect would necessarily be good to add to this one-trick pony. One could write reinit_after_move<reset_on_copy<int>> thing_ to pick up both effects (I think). If it were to do both things I think we'd have to call it something else.

I can appreciate that line of thinking, I'd also considered that.  In general, a moved-from object is in a valid-but-indeterminate state.  Nobody is allowed to ask what's in there anyway, so in one sense it doesn't matter at all what's in there (either your suggestion or mine).  (Specific types that are moveable can make a stronger guarantee, but many don't, and the ones that do tend to promise that they are "as-if-default-constructed", not "still-usable-as-if-merely-copied".)  Because a typical valid-but-indeterminate state is the default-constructed value, which in our case we happily already know, I'd say we might as well use it, especially if we expect this wrapper type to be used on (borrowed) pointers, for which the moved-from alias should _definitely_ not be re-used.  I can buy the "let the user compose the wrappers" proposal you note, but I think my advocacy is that wanting both is the 99% case (inclusive of any don't-care-one-way-or-the-other uses), and that the "keep a stale bitwise copy" is the 1% case, and dangerous in 50% of the others, so we might as well make this convenience wrapper do the most-likely-correct-and-safest thing.

jwnimmer-tri commented:
*[common/test/reinit_on_copy_test.cc, line 27 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6TcRUd-VFyaqME6u92:-L6TsKIKDr3kbCsUNChB:b-uzirfa) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L27)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

So, I feel those are two different issues:

1. I'm not persuaded that "reset" and "reinit" are any different in this regard.
2. In the example I gave above, `use_count_` does *not* get reset to 3, but to 0. That's the unexpected effect to which I was alluding.
</blockquote></details>

Maybe `reinit_on_copy(const T& value)` should not exist, but rather only `operator=(const T& value)`?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[common/test/reinit_on_copy_test.cc, line 27 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6TcRUd-VFyaqME6u92:-L6TsVCf6QUHDJrR21Vr:bexzrv1) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L27)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Maybe `reinit_on_copy(const T& value)` should not exist, but rather only `operator=(const T& value)`?
</blockquote></details>

(For an object whose job in life is to start out zero, allowing users to specify non-zero seems wrong.)

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Checkpoint -- responded to most comments, PTAL.

---

Reviewed 3 of 4 files at r1, 2 of 5 files at r2.
Review status: 3 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[common/BUILD.bazel, line 41 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6TZORqvk8S158YaCLu:-L6TnKZA6lH6hO6TpKTU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/BUILD.bazel#L41)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Please list the new item here.
</blockquote></details>

Done.

---

*[common/reinit_on_copy.h, line 4 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6TZd4a2YLihJLVRN93:-L6Tor-FDw5gC9Ug-zef:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/reinit_on_copy.h#L4)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Unused?
</blockquote></details>

Done.

---

*[common/reinit_on_copy.h, line 11 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6TZe_Q9kLZ0qoA8jhc:-L6Tp7-l3Ply_GVrayRK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/reinit_on_copy.h#L11)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I'd recommend borrowing more from the url you've linked below on the constructor documentation, e.g.:

> Note that value initialization means the initialization performed when a variable is constructed with an empty initializer. For example, numeric types are set to zero and pointer types are set to nullptr.

This leaves the door slightly more ajar to predict what happens when `T` is *not* POD.
</blockquote></details>

Done.

---

*[common/reinit_on_copy.h, line 76 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6TaNweAZ10jGtC0015:-L6TxuQm4qdNw1h_lPd-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/reinit_on_copy.h#L76)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Seems `reinit_after_move` would benefit from this as well. At the very least, it would make the two otherwise parallel APIs align better.
</blockquote></details>

Done.

---

*[common/test/reinit_on_copy_test.cc, line 13 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6Tatf3MRn7Q9OQuqSx:-L6TqM566qlUdfauyZ6Z:bqh05s3) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L13)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

It seems to me that this trick is *absolutely* necessary if `T` is non-POD.
</blockquote></details>

I'm not sure whether you are suggesting a change here, Sean or if this is just a (blocking!) side comment?

---

*[common/test/reinit_on_copy_test.cc, line 27 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6TcRUd-VFyaqME6u92:-L6Tq_z6-97UDAULAh-9:b28x5dh) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L27)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

(For an object whose job in life is to start out zero, allowing users to specify non-zero seems wrong.)
</blockquote></details>

I understood what you were saying, Sean. "reinit" implies "re-initialization" so would lead someone to think that the member would be re-initialized to whatever it was set to originally. Simbody's ReinitOnCopy does exactly that, so in your example would really get initialized to 3. The word "reset" at least doesn't have initialization in the name and at least sometimes has connotations of setting to zero. For example, `unique_ptr::reset()` nulls the pointer.  Also, in electronics reset means "set to zero", see the definition [here](https://www.google.com/search?q=define+reset).

---

*[common/test/reinit_on_copy_test.cc, line 84 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6Tblk2OF5vevt5BBYA:-L6TuKWk69zHSiV8jRmL:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L84)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Perhaps one more test?
`EXPECT_EQ(&*my_foo_ptr, &my_foo);`
The tests above could pass, even if `my_foo` were cloned in the construction. This additional test precludes the possibility of cloning.
</blockquote></details>

Done.

---

*[common/test/reinit_on_copy_test.cc, line 107 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6Tc9AszNH4ltr1VugU:-L6Txpkf6jZ0boZLnfZG:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L107)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

See previous note on why this is surprising.
</blockquote></details>

Done.

---

*[common/test/reinit_on_copy_test.cc, line 155 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6Td-7lcMv3el8WA9Bn:-L6TxxQqCXwZ6hmzt-8t:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L155)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI This value 22 is never used.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-:-L6TyFVo6WIdMLFXBaO7:b-4c7pmi)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
FYI @sherm1 because you "marked as reviewed" (green eyeball) the discussion workflow is screwed up -- when I ack things, they aren't resolved, they are blocked on you to ack.  I think if you put the eyeballs back to red it will be okay again.

---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[common/reinit_on_copy.h, line 68 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6U-eN2u8qkJCadPoRu:-L6U-eN3bpua3nNgMuwq:b-3aaat9) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1dc1ce55164a8374971b1929566d1dc44d5304/common/reinit_on_copy.h#L68)):*
> ```objc
>   /// Copy assignment value-initializes _except_ for self-assignment.
>   reinit_on_copy& operator=(const reinit_on_copy& source) {
>     if (this != &source)
> ```

FYI Is there a unit test for self-assignment?  I didn't immediately see it.

---

*[common/reinit_on_copy.h, line 69 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6U-HAY5hKG3huWYTgW:-L6U-HAY5hKG3huWYTgX:bbwj5iv) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1dc1ce55164a8374971b1929566d1dc44d5304/common/reinit_on_copy.h#L69)):*
> ```objc
>   reinit_on_copy& operator=(const reinit_on_copy& source) {
>     if (this != &source)
>       new (&value_) T{};  // Use placement new to reconstruct.
> ```

Don't we have to placement-delete first?  I suspect `value_ = {};` is just as efficient, without such a sharp tool.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6TzwkPrI8dFIww-lrB:btnocdo)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Crap -- I thought clicking those was only visible to me :(. Clicking them again makes them go to "hollow green" but I can't get them back to red. Sorry!

---

Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-:-L6U-wFt1Pgm7hWKERzB:btlq6qy)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
I pushed the rename which turned most things back to red -- hope that helped!

---

Review status: 2 of 6 files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-:-L6U0g7iD9gnty-VzgtS:b-tavi6a)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Ah yes, hollow-green would have been fine too.  Things seem fine now in any case.

---

Review status: 2 of 6 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6U18EOkdyZLLHLnFLg:bdkbw9b)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 4 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[common/test/reset_on_copy_test.cc, line 118 at r3](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6U1_c91snm4UOnzV3s:-L6U1_c91snm4UOnzV3t:b1fcaac) ([raw file](https://github.com/robotlocomotion/drake/blob/bf85493766766509291fe2c9613050c51b03b417/common/test/reset_on_copy_test.cc#L118)):*
> ```cpp
>   // Self copy assignment preserves value.
>   w = w;
>   EXPECT_EQ(w, 4);
> ```

@jwnimmer-tri, self-assignment check is here.

---

*[common/reinit_on_copy.h, line 68 at r2](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6U-eN2u8qkJCadPoRu:-L6U1m1C10Yjn6xmK1HY:b-o9k7gu) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1dc1ce55164a8374971b1929566d1dc44d5304/common/reinit_on_copy.h#L68)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Is there a unit test for self-assignment?  I didn't immediately see it.
</blockquote></details>

OK, line 118 (I added a note there in case it moves).

---

*[common/reinit_on_copy.h, line 69 at r2](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6U-HAY5hKG3huWYTgW:-L6U3XxqEOfnoL06kx2U:baards9) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1dc1ce55164a8374971b1929566d1dc44d5304/common/reinit_on_copy.h#L69)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Don't we have to placement-delete first?  I suspect `value_ = {};` is just as efficient, without such a sharp tool.
</blockquote></details>

Yikes -- I tried to be too macho! Yes you're right, thanks -- fixed with softer semantics.

---

*[common/test/reinit_on_copy_test.cc, line 29 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6TaistWLjR0eXWs1oj:-L6UAStMEyTJ3_pcWXHC:bt2gbmo) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reinit_on_copy_test.cc#L29)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Adding a non-POD example might help illustrate some of the list intuitive aspects:

```C++
struct Bar {
  DRAKE_DEFAULT_COPY_AND_MOVE_AND_ASSIGN(Bar)
  Bar() = default;
  int i{10};
};

int& BarI(Bar& b) {
  return b.i;
}

GTEST_TEST(ReinitOnCopyTest, NonPodData) {
  reinit_on_copy<Bar> b;
  EXPECT_EQ(BarI(b), 10);
  BarI(b) = 15;
  EXPECT_EQ(BarI(b), 15);
  reinit_on_copy<Bar>b_cpy(b);
  // Note: the *contained* numerical values are value-initialized to the declared default.
  EXPECT_EQ(BarI(b_cpy), 10);
}
```
</blockquote></details>

Done. Modified it some, PTAL.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 6 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@jwnimmer-tri, I have to say I do like the idea of thowing in the reinit_after_move functionality here also, especially since we already know that type T can be value-initialized. So I made that change.

I added a test case to show how std::vector behaves with move vs copy. When move resets like copy, std::vector's resize method wipes out the existing elements. Admittedly, unless you say the right incantations C++ will use copy construction rather than move construction, and anyone who uses reset_on_copy on complicated objects better think carefully (I added a note to that effect). However, I think it will be used mostly for ints and pointers and it would be nice if `std::vector<C>` doesn't misbehave just because C has an otherwise-harmless `reset_on_copy<int>` member.
Making this work properly did take some additional incantations here, mostly involving noexcept. PTAL

I'll fix up reinit_after_move to match after we settle on how reset_on_copy should behave.

---

Review status: 4 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[common/reinit_on_copy.h, line 11 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-L6TZe_Q9kLZ0qoA8jhc:-L6U1-TN5y9dJDsi4DA6:bfftovk) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/reinit_on_copy.h#L11)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Also see `@note` added below.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8208#-:-L6VLreECx1Qr1wtbdQr:bojzk4n)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  I also tested locally that `reset_on_copy<const int>` (when used) and `reset_on_copy<int&>` (when declared) give suitably scary compiler errors.

> I added a test case to show how std::vector ...

Aha!  You were talking about if someone did a `vector<reset_on_copy<Foo>>` and you don't want the Foos to get reset just because the vector capacity changed.  That is very clear thanks, and I agree it would be surprising if vector capacity changes caused resets.

+(status: curate commits before merging) 

---

Reviewed 2 of 6 files at r3, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6WokXwKBFpobo2eyYF:-L6WokXwKBFpobo2eyYG:bglunm6) (no related file):*
BTW I think the reasons to use `reset` over `reinit` apply equally well to reinit_after_move; I think a subsequent PR should rename that one also, for consistency?

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6WolidL1kYEtDTgTr-:-L6WolidL1kYEtDTgTr0:b-qi0xel) (no related file):*
BTW The PR title probably needs a rename now also.

---

*[common/reset_on_copy.h, line 90 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Wkj_X4eYjTD5IJuIl:-L6Wkj_X4eYjTD5IJuIm:b2es2yp) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L90)):*
> ```objc
>   /// The source argument is otherwise ignored.
>   reset_on_copy& operator=(const reset_on_copy& source) noexcept(
>   std::is_nothrow_destructible<T>::value &&
> ```

BTW This indentation seems wrong.

---

*[common/reset_on_copy.h, line 154 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6WmXPsplsfstzA9KBz:-L6WmXPsplsfstzA9KC-:bh4sov9) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L154)):*
> ```objc
> 
>  private:
>   // Invokes T's destructor if there is one, then value-initializes.
> ```

FYI Probably worth documenting that we use this implementation tactic so that only the ctor and dtor have to be noexcept -- that the more obvious `value_ = T{};'` would require move-assign to be noexcept.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6WjVg9PCkH7X2EE2Iu:b-vvf5th)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with some last thoughts:

1. Some cosmetic comments below
2. When `T` is a struct/class, there are some non-obvious implications that are not well documented (only implicitly in the tests).
    1. The importance of a `noexcept` move constructor for use in an `std::vector`.
    2. The `->` operator to reference members of `T` (rather than the `.` operator).

I think these subtleties should be explicitly called out in the class documentation in a "Working with non-POD data" kind of section.

---

Reviewed 2 of 7 files at r2, 4 of 6 files at r3, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[common/reset_on_copy.h, line 79 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Wz2aTrlAqZ8gLUl1T:-L6Wz2aTrlAqZ8gLUl1U:b-tp1nxv) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L79)):*
> ```objc
>   /// @name Implements copy/move construction/assignment.
>   /// These make %reset_on_copy objects CopyConstructible, CopyAssignable,
>   /// MoveConstructible, and MoveAssignable. These methods will be noexcept
> ```

FYI This comment (re: `noexcept`) would also apply to the constructors (but it isn't explicitly called out there).

---

*[common/reset_on_copy.h, line 82 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Wz6MuTd48sr855_4Z:-L6Wz6MuTd48sr855_4_:bdhfhj7) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L82)):*
> ```objc
>   /// if the required methods of type T are noexcept.
>   //@{
>   /// Copy constructor just value-initializes instead; the source is ignored.
> ```

FYI I would prefer some whitespace between the group's documentation and the method's documentation; otherwise, the eye would seek to interpret the *full* comment block as method docs. That said, I recognize there is no white space between groups anywhere in this file, or the `reset_on_move` origin either.

And, somewhat whimsically, I'd *almost* like to see the word "Copy" in quotes. :)

---

*[common/test/reset_on_copy_test.cc, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Tatf3MRn7Q9OQuqSx:-L6X-e1iQpA7CoYHzNwH:b7p5yml) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reset_on_copy_test.cc#L13)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I'm not sure whether you are suggesting a change here, Sean or if this is just a (blocking!) side comment?
</blockquote></details>

My original comment was rendered moot by the highly surprising use of the `->` operator (see other notes).  I.e., the contrast between these two things:

```C++
int i;
reset_on_copy<int> i_roc;

i = 10;        // looks like int-to-int assignment
i_roc = 10;    // looks like int-to-int assignment.

Bar b;
reset_on_copy<Bar> b_roc;
b.i = 10;       // looks like assignment to _object_ member assignment
b_roc->i = 10;  // looks like assignment to _pointer_ member assignment.
b_roc.i = 10;   // This doesn't compile.
```

---

*[common/test/reset_on_copy_test.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6TcRUd-VFyaqME6u92:-L6X0NgKbXmZIStU6BYB:bz3t0oq) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reset_on_copy_test.cc#L27)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I understood what you were saying, Sean. "reinit" implies "re-initialization" so would lead someone to think that the member would be re-initialized to whatever it was set to originally. Simbody's ReinitOnCopy does exactly that, so in your example would really get initialized to 3. The word "reset" at least doesn't have initialization in the name and at least sometimes has connotations of setting to zero. For example, `unique_ptr::reset()` nulls the pointer.  Also, in electronics reset means "set to zero", see the definition [here](https://www.google.com/search?q=define+reset).
</blockquote></details>

The question: does the name `reset` sufficiently inform users of the unexpected behavior? Or should we call that out in the documentation so they are not required to infer it?

---

*[common/test/reset_on_copy_test.cc, line 262 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6X1tqkLIelaeDYypjT:-L6X1tqkLIelaeDYypjU:b-1siz1r) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/test/reset_on_copy_test.cc#L262)):*
> ```cpp
> }
> 
> // std::vector when resizing has to move elements from the old memory to
> ```

FYI Good catch. It's amazing how the subtle nuances of this kind of thing spirals like this.

---

*[common/test/reset_on_copy_test.cc, line 288 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6X2BiEVkhq1tZihDXr:-L6X2BiFay-m3_sgd3J2:b-yuh7wh) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/test/reset_on_copy_test.cc#L288)):*
> ```cpp
>   std::vector<reset_on_copy<CanMove>> v;
>   v.emplace_back(CanMove(1)); v.emplace_back(CanMove(2));
>   EXPECT_EQ(v[0]->i, 1);
> ```

FYI *This* strikes me as a counter-intuitive operation. It gives the `reset_on_copy` pointer-like semantics. I think this should be added to the class's main documentation in a section dedicated to the weird implications of non-POD `T` types.

---

*[common/test/reset_on_copy_test.cc, line 291 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6X2a_8fR9QPj54eHd3:-L6X2a_8fR9QPj54eHd4:b-d0btvs) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/test/reset_on_copy_test.cc#L291)):*
> ```cpp
>   EXPECT_EQ(v[1]->i, 2);
>   const size_t old_v_capacity = v.capacity();
>   v.resize(100);
> ```

FYI Wouldn't it make sense to resize it to `old_v_capacity + delta`? Guarantees that new capacity and old capacity *can't* be equal.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6X3b7RCHCHh7jpPu-8:b-cb5wp6)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Another option would be to restrict `reset_on_copy` to require that T is `std::is_fundamental`.  Then a lot of this complexity just disappears.

SeanCurtis-TRI commented:
I'd be content with that.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6XA1ATdviH3hW0is5R:bbc9xgy)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Me too. I only have use cases at the moment for integers and pointers. Probably `std::is_scalar` would be a good restriction (leaves out `void` and `nullptr_t`). Enums are scalars but resetting them is _slightly_ sketchy since they get reset to zero even if zero is not one of the spelled-out enumeration values. I could exclude them but I suppose enums are a bit sketchy anyway.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6WokXwKBFpobo2eyYF:-L6XYaVofD2TH3TdepOc:b-vy7l0z) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW I think the reasons to use `reset` over `reinit` apply equally well to reinit_after_move; I think a subsequent PR should rename that one also, for consistency?
</blockquote></details>

Agreed. reset_after_move would be better. I'll fix in followup pr.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6WolidL1kYEtDTgTr-:-L6XZ0BvSzSixhv4k1jO:bkypgs3) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW The PR title probably needs a rename now also.
</blockquote></details>

Done (PR retitled).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6XXdIaVMApNhkNBP9U:b3shxf6)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Actually, it still leaves `std::nullptr_t` in the working set. But it seems that `is_scalar` is the most targeted limiting set.

---

Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6Y4kQipfLN2iMz5Pqo:byma643)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Oh yeah, I missed that. Well I guess resetting a nullptr shouldn't present much of a problem!

---

Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6YIjBq-C6v7vEqHWrw:b-d273z1)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Simplified class by restricting to scalar types, all reviewer comments resolved or responded. PTAL in case you have any final comments.

---

Review status: 3 of 6 files reviewed at latest revision, 9 unresolved discussions.

---

*[common/reset_on_copy.h, line 79 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Wz2aTrlAqZ8gLUl1T:-L6YaPUV5xsi7IASQE_y:b-8p7w6y) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L79)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI This comment (re: `noexcept`) would also apply to the constructors (but it isn't explicitly called out there).
</blockquote></details>

Done. I removed the discussion and replaced it with a single 

---

*[common/reset_on_copy.h, line 82 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Wz6MuTd48sr855_4Z:-L6YcTpIS8D6Au6pF-oN:bhtq6p7) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L82)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI I would prefer some whitespace between the group's documentation and the method's documentation; otherwise, the eye would seek to interpret the *full* comment block as method docs. That said, I recognize there is no white space between groups anywhere in this file, or the `reset_on_move` origin either.

And, somewhat whimsically, I'd *almost* like to see the word "Copy" in quotes. :)
</blockquote></details>

Done. (Except for the whimsy -- it's still a copy constructor, though with some "non traditional" semantics.)

---

*[common/reset_on_copy.h, line 90 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Wkj_X4eYjTD5IJuIl:-L6Ycoe1KBqt_161FBoh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L90)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW This indentation seems wrong.
</blockquote></details>

Done.

---

*[common/reset_on_copy.h, line 154 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6WmXPsplsfstzA9KBz:-L6YcwDPJdCIDzuxmZkP:bxda1qx) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L154)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Probably worth documenting that we use this implementation tactic so that only the ctor and dtor have to be noexcept -- that the more obvious `value_ = T{};'` would require move-assign to be noexcept.
</blockquote></details>

Done. With the is_scalar restriction we can punt this discussion to the far future since everything is noexcept.

---

*[common/test/reset_on_copy_test.cc, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Tatf3MRn7Q9OQuqSx:-L6Yd7xnF_xbN2XqZpQF:b-chfu2k) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reset_on_copy_test.cc#L13)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

My original comment was rendered moot by the highly surprising use of the `->` operator (see other notes).  I.e., the contrast between these two things:

```C++
int i;
reset_on_copy<int> i_roc;

i = 10;        // looks like int-to-int assignment
i_roc = 10;    // looks like int-to-int assignment.

Bar b;
reset_on_copy<Bar> b_roc;
b.i = 10;       // looks like assignment to _object_ member assignment
b_roc->i = 10;  // looks like assignment to _pointer_ member assignment.
b_roc.i = 10;   // This doesn't compile.
```
</blockquote></details>

Done. Got rid of the weird operators and restricted to is_scalar<T>.

---

*[common/test/reset_on_copy_test.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6TcRUd-VFyaqME6u92:-L6YdHHMyz6T267ARAzU:b8u2ffz) ([raw file](https://github.com/robotlocomotion/drake/blob/d66d5f7da88ead103b6150400fbbd974a6c668b9/common/test/reset_on_copy_test.cc#L27)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

The question: does the name `reset` sufficiently inform users of the unexpected behavior? Or should we call that out in the documentation so they are not required to infer it?
</blockquote></details>

Done. Added a warning to the class description, PTAL.

---

*[common/test/reset_on_copy_test.cc, line 262 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6X1tqkLIelaeDYypjT:-L6YeRLYLZDmHQ2i6wOb:bo6594o) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/test/reset_on_copy_test.cc#L262)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Good catch. It's amazing how the subtle nuances of this kind of thing spirals like this.
</blockquote></details>

Done. Moot now.

---

*[common/test/reset_on_copy_test.cc, line 288 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6X2BiEVkhq1tZihDXr:-L6YeS_5edmNm_DAsR8N:bo6594o) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/test/reset_on_copy_test.cc#L288)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI *This* strikes me as a counter-intuitive operation. It gives the `reset_on_copy` pointer-like semantics. I think this should be added to the class's main documentation in a section dedicated to the weird implications of non-POD `T` types.
</blockquote></details>

Done. Moot now.

---

*[common/test/reset_on_copy_test.cc, line 291 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6X2a_8fR9QPj54eHd3:-L6YeO6KPKD3A0sGGi84:bo6594o) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/test/reset_on_copy_test.cc#L291)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Wouldn't it make sense to resize it to `old_v_capacity + delta`? Guarantees that new capacity and old capacity *can't* be equal.
</blockquote></details>

Done. Moot now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6Yeq7NffCNj2I9LG8R:brbyp68)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 6 files at r3, 3 of 3 files at r6.
Review status: 5 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[common/reset_on_copy.h, line 90 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Wkj_X4eYjTD5IJuIl:-L6YfgSiro-JmrAh0syF:btbss9r) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L90)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

No change?  Is clang-format putting it back this way?

---

*[common/reset_on_copy.h, line 154 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6WmXPsplsfstzA9KBz:-L6YfFaZwhtfVviSRQK7:b-nz8vy) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L154)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Done. With the is_scalar restriction we can punt this discussion to the far future since everything is noexcept.
</blockquote></details>

Well, the _reason_ we're doing the placement hijinks is because we need to be careful about which operations are noexcept, yes?  A chesterton's comment along the lines of "Don't change this to `value_ = T{};` because..." might help here, was my point.

---

*[common/reset_on_copy.h, line 79 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Yf-ByNYqIHNOHlpoD:-L6Yf-Bzuq6Vag221Id2:bn9emyy) ([raw file](https://github.com/robotlocomotion/drake/blob/6e8f2e9e7ec72437d6cd8294b364971ddf157b24/common/reset_on_copy.h#L79)):*
> ```objc
>                 "reset_on_copy<T> is permitted only for integral, "
>                 "floating point, and pointer types T.");
>  public:
> ```

FYI GSG wants all public / private / etc. in one spot.  I think putting the public prior to this line is an easy fix.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 5 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[common/reset_on_copy.h, line 90 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Wkj_X4eYjTD5IJuIl:-L6YgMMXIcW3d6T19Z7r:bwt5wxm) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L90)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

No change?  Is clang-format putting it back this way?
</blockquote></details>

Odd -- I made the change and I'm seeing it in Reviewable. Are you looking at the lastest? (I rebased so Reviewable may have stumbled.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 2 of 6 files at r3, 2 of 3 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[common/reset_on_copy.h, line 90 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Wkj_X4eYjTD5IJuIl:-L6YgR9U2rDr68JtmEkB:b-q59rpr) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L90)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

No change?  Is clang-format putting it back this way?
</blockquote></details>

Done now.

---

*[common/reset_on_copy.h, line 18 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6YfljEhfSUvYIKTfwi:-L6YfljEhfSUvYIKTfwj:bmz35ug) ([raw file](https://github.com/robotlocomotion/drake/blob/7f4ef2337b832fa76fa6413ccd8b0f2308a47e0e/common/reset_on_copy.h#L18)):*
> ```objc
> /// Only types T that satisfy `std::is_scalar<T>` are currently
> /// permitted: integral and floating point types, enums, and pointers.
> /// Value initialization means the initialization performed when a variable is
> ```

BTW This sentence is no longer required -- it's an artifact of non-scalar `T`. The following sentence is sufficient.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[common/reset_on_copy.h, line 90 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Wkj_X4eYjTD5IJuIl:-L6Yh0F8hAK-RrZJcZZC:bo0mnxq) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L90)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Done now.
</blockquote></details>

OK Looks good now.  I think I had a PEBKAC moment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Done now? PTAL.

---

Review status: 5 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[common/reset_on_copy.h, line 154 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6WmXPsplsfstzA9KBz:-L6YiKadXIaz-NmjxiqR:b4n8kv9) ([raw file](https://github.com/robotlocomotion/drake/blob/5afaa5ed111eafb33feea31c5bbb19ce8c94370f/common/reset_on_copy.h#L154)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Well, the _reason_ we're doing the placement hijinks is because we need to be careful about which operations are noexcept, yes?  A chesterton's comment along the lines of "Don't change this to `value_ = T{};` because..." might help here, was my point.
</blockquote></details>

Done. Gotcha -- misunderstood before, ptal.

---

*[common/reset_on_copy.h, line 79 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6Yf-ByNYqIHNOHlpoD:-L6YivgCi_HPe71vZE-C:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6e8f2e9e7ec72437d6cd8294b364971ddf157b24/common/reset_on_copy.h#L79)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI GSG wants all public / private / etc. in one spot.  I think putting the public prior to this line is an easy fix.
</blockquote></details>

Done.

---

*[common/reset_on_copy.h, line 18 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6YfljEhfSUvYIKTfwi:-L6Yi_nSJniSpXmBGIaI:bbr2xia) ([raw file](https://github.com/robotlocomotion/drake/blob/7f4ef2337b832fa76fa6413ccd8b0f2308a47e0e/common/reset_on_copy.h#L18)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW This sentence is no longer required -- it's an artifact of non-scalar `T`. The following sentence is sufficient.
</blockquote></details>

OK, I think it still applies here even with scalar types. For example, `int i{};` is an empty initializer and sets i to zero.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6YjduwxrifQ7rbS70F:b6zna2p)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 6 files at r3, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
I believe I'm out of the way.

---

Reviewed 1 of 6 files at r3, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[common/reset_on_copy.h, line 18 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6YfljEhfSUvYIKTfwi:-L6YkOyfr5esVIs-LAE0:b-biwpan) ([raw file](https://github.com/robotlocomotion/drake/blob/7f4ef2337b832fa76fa6413ccd8b0f2308a47e0e/common/reset_on_copy.h#L18)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

OK, I think it still applies here even with scalar types. For example, `int i{};` is an empty initializer and sets i to zero.
</blockquote></details>

I won't block -- but the idea of "general" value initialization is mostly useless. Particularly when you provide a very specific, and largely all-encompassing, definition in the next sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6Ykg4QuTN8uEz8Pwm1:b6ga6vg)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Passed CI. Just a comment fix after that so I'll force merge.

---

Review status: 5 of 6 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[common/reset_on_copy.h, line 18 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-L6YfljEhfSUvYIKTfwi:-L6YzB-LnHDxpM_lKKjV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7f4ef2337b832fa76fa6413ccd8b0f2308a47e0e/common/reset_on_copy.h#L18)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I won't block -- but the idea of "general" value initialization is mostly useless. Particularly when you provide a very specific, and largely all-encompassing, definition in the next sentence.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8208#-:-L6YzDUSe52zJ4k_1IqQ:b-5vl2hq)*
<!-- Sent from Reviewable.io -->
